prime-router/docs/proposals/lineage/README.md

```diff
@@ -0,0 +1,123 @@
+- A few schema best practices:  Every table must have a single col PK (either a serial or a UUID), and a created_date, Normalize independent data.   Use singular nouns.
+-- Receipt of a dataset implies the right to see the lineage of that dataset.
+-- Sending a dataset implies the right to see the lineage of that dataset. 
+
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
+(Note: [Original lucid chart diagram is here](https://lucid.app/lucidchart/7b81cccb-91a0-44b3-b17f-90b9b0c8304e/edit?beaconFlowId=332110D5160D6847&page=0_0#))
+
+### Notes on the Proposed Schema
+
+- The separation of Task from Report allows us to have Tasks that don't necessarily create Reports.
+- Addition of the `TASK.task_result` fills a gap where we're not tracking the error messages sent back to users.
+- Apparently I can't name the REPORT table 'REPORT', due to a limitation in Flyway/jooq.  Hence the weird name `REPORT_FILE`.   I also thought about `REPORT_IN`, `REPORT_DATA`, `REPORT_REC`.  Haven't sunk as low as `REPURT`, but I'm on my way there...
+- Many values in REPORT_FILE will be empty depending on the task creating the report.  For example, all the sending_* fields are only filled in for an initially received report.   All the receiving_* fields are only filled in for reports sent externally.
+- might need a report_status in the future.  For example, `wiped` or `archived`, to explain why a blob might have disappeared.
+- The primary key to use for ITEM is tricky.  The sender gives us the `tracking_element`, but we have little control over how robust that is.  Because of this, all queries on `tracking_element` must also filter on report_id.
+- So there's an open question:  should we require global uniqueness of `tracking_elements`?   It would be wonderfully convenient to use `tracking_elements` as primary keys in our database.    However, they would have to be globally unique, because report data can get merged together in arbitrary ways in steps downstream from Receive.
+- There's subtlety with Report parent/child relationships.   Because any given task might take in a bunch of reports, and produce a bunch of reports, there's no guarantee that any of the data in output Report Y came from input Report X.  However we're tracking it in the REPORT_LINEAGE table as if every input report fed every output report.  In practice, its very hard to tell if that's true, in general.
+
+### How this design helps build trust
+
+- Fine-grained lineage tracking at the ITEM level, with ability to query both "backward" and "forward"
+- Improved tracking of sender and receiver ownership
+- Tracking versions of the functions that created the data, and SHA hash of every report.
+- Next steps: Have an easy-to-use API that delivers history
+- Next steps: Have an easy-to-use API that can retrieve the data in an individual ITEM. (?)
+
+### Rollout plan:
+
```

Comment:
When you say rollout plan, I assuming that you talking about separate PRs that could go to production? I wonder what the operational implications of having two sets of tables going in parallel means? I suggest that the two at once state will be somewhat problematic. If so, could we simplify this plan into fewer steps that executed faster? We can take advantage of the fact that we have very low transaction rate now.

Reply:
I plan to write the code to populate the new model without touching the old one.    At some point, that's done, and we can decide whether to deploy to production, or wait until the old style is removed.

Personally, I like the low risk of running in parallel, for a little while.   As long as nothing in production is reading the new tables, they are completely harmless, even if they are erroneous in some way.